src/sciline/_provider.py

```diff
@@ -259,7 +274,46 @@ def _module_name(x: Any) -> str:
     return getattr(inspect.getmodule(x), '__name__', '')
 
 
-def _get_provider_args(func: ToProvider) -> inspect.FullArgSpec:
+def _unwrap_decorated(func: ToProvider) -> ToProvider:
     if (wrapped := getattr(func, '__wrapped__', None)) is not None:
-        return _get_provider_args(wrapped)
-    return inspect.getfullargspec(func)
+        return _unwrap_decorated(wrapped)
+    return func
+
+
+def _get_args_and_types(
+    func: ToProvider,
+) -> tuple[list[str], list[str], dict[str, Any]]:
+    if inspect.isfunction(func):
+        return _get_func_args_and_types(func)
+    if inspect.ismethod(func):
+        return _get_method_args_and_types(func)
+    if isinstance(func, type):
+        raise TypeError(
+            "Classes are not supported as providers because their type annotations "
+            f"cannot reliably be extracted. Got class {func}. "
+            "Consider using a free function, classmethod, or staticmethod."
+        ) from None
+    if (call := getattr(func, '__call__', None)) is not None:  # noqa: B004
+        return _get_method_args_and_types(call)
+    raise TypeError(f"Cannot get type hints of provider {func}") from None
+
+
+def _get_func_args_and_types(
+    func: ToProvider,
+) -> tuple[list[str], list[str], dict[str, Any]]:
+    func = _unwrap_decorated(func)
+    hints = get_type_hints(func)
+    signature = inspect.getfullargspec(func)
+    return signature.args, signature.kwonlyargs, hints
+
+
+def _get_method_args_and_types(
+    func: ToProvider,
+) -> tuple[list[str], list[str], dict[str, Any]]:
+    func = _unwrap_decorated(func)
+    args, kwonlyargs, hints = _get_func_args_and_types(func)
+    # Remove annotation of `self` / `cls` if present
+    hints.pop(args[0], None)
+    # Drop `self` / `cls`
+    args = args[1:]
```

Comment:
What if this is a `staticmethod`? Does this actually take the `inspect.isfunction` branch?

Reply:
Yes. See https://github.com/scipp/sciline/blob/1a75b95cef64887bdeeed8fd16899fcc20044192/tests/pipeline_test.py#L1507